File: src/main/java/org/noxet/noxetserver/menus/inventory/HomeNavigationMenu.java
package org.noxet.noxetserver.menus.inventory;

import org.bukkit.DyeColor;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.block.banner.Pattern;
import org.bukkit.block.banner.PatternType;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.ClickType;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.BannerMeta;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;
import org.noxet.noxetserver.NoxetServer;
import org.noxet.noxetserver.RealmManager;
import org.noxet.noxetserver.UsernameStorageManager;
import org.noxet.noxetserver.commands.misc.Friend;
import org.noxet.noxetserver.commands.teleportation.Home;
import org.noxet.noxetserver.menus.ItemGenerator;
import org.noxet.noxetserver.menus.chat.ChatPromptMenu;
import org.noxet.noxetserver.playerdata.PlayerDataManager;
import org.noxet.noxetserver.util.InventoryCoordinate;
import org.noxet.noxetserver.util.InventoryCoordinateUtil;

import java.text.DecimalFormat;
import java.util.*;

public class HomeNavigationMenu extends InventoryMenu {
    private final Player player;
    private final RealmManager.Realm realm;
    private final Map<String, Location> homes;
    private final InventoryCoordinate toggleFriendsHomesSlot;
    private final boolean showFriendsHomes;
    private final List<String> friendsHomes;

    private final BukkitTask toggleFriendsHomesSlotUpdateTimer;
    private int friendSkullScrollIndex = 0;

    public HomeNavigationMenu(Player player, RealmManager.Realm realm) {
        super(
                (Home.getRealmHomes(player, realm).size() - 1) / 9 + 2 +
                ((boolean) new PlayerDataManager(player).get(PlayerDataManager.Attribute.SHOW_FRIEND_HOMES) ? (Home.getFriendsHomes(player, realm).size() - 1) / 9 : 0),
                "⚑ " + realm.getDisplayName() + " Homes",
                false
        );

        homes = Home.getRealmHomes(player, realm);
        this.player = player;
        this.realm = realm;

        showFriendsHomes = (boolean) new PlayerDataManager(player).get(PlayerDataManager.Attribute.SHOW_FRIEND_HOMES);

        toggleFriendsHomesSlot = InventoryCoordinateUtil.getCoordinateFromXY(0, getInventory().getSize() / 9 - 1);

        friendsHomes = showFriendsHomes ? Home.getFriendsHomes(player, realm) : null;

        List<String> friends = Friend.getFriendList(player.getUniqueId());

        toggleFriendsHomesSlotUpdateTimer = new BukkitRunnable() {
            @Override
            public void run() {
                if(friends.size() > 0)
                    friendSkullScrollIndex %= friends.size();

                setSlotItem(
                        ItemGenerator.generatePlayerSkull(
                                friends.size() > 0 ? NoxetServer.getPlugin().getServer().getOfflinePlayer(
                                        new UsernameStorageManager().getUUIDFromUsernameOrUUID(
                                                friends.get(friendSkullScrollIndex++)
                                        )
                                ) : player,
                                "§aToggle Friends' Homes",
                                Arrays.asList(
                                        showFriendsHomes ? "§a§l✔ SHOWN" : "§c§l❌ HIDDEN",
                                        "§7When enabled, your friends' shared",
                                        "§7homes will be shown in this list.",
                                        "§e→ Click to " + (showFriendsHomes ? "hide" : "show") + "."
                                )
                        ), toggleFriendsHomesSlot
                );
            }
        }.runTaskTimer(NoxetServer.getPlugin(), 0, 40);
    }

    @Override
    protected void updateInventory() {
        int i = 0;

        for(Map.Entry<String, Location> home : homes.entrySet()) {
            Location homeLocation = home.getValue();
            World.Environment dimension = Objects.requireNonNull(homeLocation.getWorld()).getEnvironment();
            String locationNote = "§5";

            switch(dimension) {
                case NORMAL:
                    locationNote += "Overworld";
                    break;
                case NETHER:
                    locationNote += "Nether";
                    break;
                case THE_END:
                    locationNote += "The End";
                    break;
                default:
                    locationNote += "???";
            }

            if(homeLocation.getWorld() == player.getWorld()) {
                // Same world: show distance.
                int blocksAway = (int) player.getLocation().distance(homeLocation);
                locationNote += " §e" + new DecimalFormat("#,###").format(blocksAway) + " blocks away.";
            }

            boolean isHomeMain = home.getKey().equals(Home.defaultHomeName);

            List<String> lore = new ArrayList<>(Arrays.asList(
                    locationNote,
                    "§8X §7" + ((int) homeLocation.getX()),
                    "§8Y §7" + ((int) homeLocation.getY()),
                    "§8Z §7" + ((int) homeLocation.getZ()),
                    "§e→ Double-click to §5§nteleport§e.",
                    "§e→ Right-click to §b§nrename§e.",
                    "§e→ Press any number on keyboard to §c§nremove§e this home.",
                    "§e→ Shift-right-click to §d§n" + (!Home.isHomeFriendShared(home.getKey()) ? "enable" : "disable") + " friend sharing§e."
            ));

            ItemStack homeItem = ItemGenerator.generateItem(
                    !isHomeMain ? Material.WHITE_BANNER : Material.RED_BED,
                    !Home.isHomeFriendShared(home.getKey()) ? (!isHomeMain ? "§a" : "§5◆ §3") + home.getKey() : "§5☮ §a" + home.getKey().substring(1),
                    lore
            );

            if(!isHomeMain)
                generateBannerPatterns(homeItem, home.getValue().hashCode());

            setSlotItem(homeItem, InventoryCoordinateUtil.getCoordinateFromSlotIndex(i++));
        }

        if(!showFriendsHomes)
            return;

        for(String friendTpId : friendsHomes) {
            // friendTpId is formatted as: player/home-name

            String playerName = friendTpId.substring(0, friendTpId.indexOf('/'));
            UUID playerUUID = new UsernameStorageManager().getUUIDFromUsernameOrUUID(playerName);

            setSlotItem(
                    ItemGenerator.generatePlayerSkull(NoxetServer.getPlugin().getServer().getOfflinePlayer(playerUUID),
                    "§3" + friendTpId,
                    Arrays.asList(
                            "§7This home is shared by §b" + playerName + "§7.",
                            "§e→ Double-click to §5§nteleport§e."
                    )), InventoryCoordinateUtil.getCoordinateFromSlotIndex(i++)
            );
        }
    }

    @Override
    protected boolean onSlotClick(Player player, InventoryCoordinate coordinate, ClickType clickType) {
        if(coordinate.isAt(toggleFriendsHomesSlot)) {
            new PlayerDataManager(player).toggleBoolean(PlayerDataManager.Attribute.SHOW_FRIEND_HOMES).save();
            new HomeNavigationMenu(player, realm).openInventory(player);
            return false;
        }

        int i = 0;

        for(Map.Entry<String, Location> home : homes.entrySet())
            if(InventoryCoordinateUtil.getCoordinateFromSlotIndex(i++).isAt(coordinate)) {
                switch(clickType) {
                    case DOUBLE_CLICK:
                        player.performCommand("home tp " + home.getKey());
                        break;
                    case NUMBER_KEY:
                        new ConfirmationMenu("Delete home '" + home.getKey() + "'?", () -> {
                            player.performCommand("home remove " + home.getKey());
                            new HomeNavigationMenu(player, realm).openInventory(player);
                        }, () -> new HomeNavigationMenu(player, realm).openInventory(player)).openInventory(player);
                        break;
                    case RIGHT:
                        new ChatPromptMenu(
                                "new name for home '" + home.getKey() + "'",
                                player,
                                promptResponse -> {
                                    String newName = promptResponse.getMessage();

                                    player.performCommand("home rename " + home.getKey() + " " + newName);
                                    new HomeNavigationMenu(player, realm).openInventory(player);
                                }
                        );
                        break;
                    case SHIFT_RIGHT:
                        boolean isAlreadyFriendShared = Home.isHomeFriendShared(home.getKey());
                        new ConfirmationMenu((isAlreadyFriendShared ? "Disable" : "Enable") + " friend sharing for home '" + home.getKey() + "'?", () -> {
                            player.performCommand("home rename " + home.getKey() + " " + (isAlreadyFriendShared ? home.getKey().substring(1) : "*" + home.getKey()));
                            new HomeNavigationMenu(player, realm).openInventory(player);
                        }, () -> new HomeNavigationMenu(player, realm).openInventory(player)).openInventory(player);
                        break;
                    default:
                        return false;
                }

                return true;
            }

        if(!showFriendsHomes)
            return false;

        for(String friendTpId : friendsHomes)
            if(InventoryCoordinateUtil.getCoordinateFromSlotIndex(i++).isAt(coordinate)) {
                switch(clickType) {
                    case DOUBLE_CLICK:
                        player.performCommand("home friend-tp " + friendTpId);
                        break;
                    default:
                        return false;
                }

                return true;
            }

        return false;
    }

    private void generateBannerPatterns(ItemStack banner, int seed) {
        Random random = new Random(seed);

        List<Pattern> patterns = new ArrayList<>();

        for(int i = 0; i < 10; i++) {
            // Append random patterns to the banner.
            // (based on location seed)
            Pattern pattern = new Pattern(
                    DyeColor.values()[random.nextInt(DyeColor.values().length)],
                    PatternType.values()[random.nextInt(PatternType.values().length)]
                    );
            patterns.add(pattern);
        }

        BannerMeta bannerMeta = (BannerMeta) banner.getItemMeta();
        assert bannerMeta != null;

        bannerMeta.setPatterns(patterns);
        bannerMeta.addItemFlags(ItemFlag.HIDE_POTION_EFFECTS); // Hide banner pattern lore.

        banner.setItemMeta(bannerMeta);
    }

    @Override
    protected void stop() {
        super.stop();

        toggleFriendsHomesSlotUpdateTimer.cancel();
    }
}

File: src/main/java/org/noxet/noxetserver/menus/inventory/HomeNavigationMenu.scala
package org.noxet.noxetserver.menus.inventory

import org.bukkit.block.banner.{Pattern, PatternType}
import org.bukkit.entity.Player
import org.bukkit.event.inventory.ClickType
import org.bukkit.inventory.meta.BannerMeta
import org.bukkit.inventory.{ItemFlag, ItemStack}
import org.bukkit.scheduler.{BukkitRunnable, BukkitTask}
import org.bukkit.{DyeColor, Location, Material, World}
import org.noxet.noxetserver.NoxetServer
import org.noxet.noxetserver.commands.social.Friend
import org.noxet.noxetserver.commands.teleportation.Home
import org.noxet.noxetserver.menus.ItemGenerator
import org.noxet.noxetserver.menus.chat.ChatPromptMenu
import org.noxet.noxetserver.playerdata.PlayerDataManager
import org.noxet.noxetserver.realm.RealmManager.Realm
import org.noxet.noxetserver.util.{InventoryCoordinate, Sample, UsernameStorageManager}

import java.text.DecimalFormat
import scala.language.postfixOps

class HomeNavigationMenu(player: Player, realm: Realm) extends InventoryMenu(
  (Home.getRealmHomes(player, realm).size - 1) / 9 + 2 +
    (if PlayerDataManager(player).get(PlayerDataManager.Attribute.SHOW_FRIEND_HOMES).toBoolean then (Home.getFriendsHomes(player, realm).size - 1) / 9 else 0),
  s"⚑ ${realm.getDisplayName} Homes",
  false):
  private val homes = Home.getRealmHomes(player, realm)
  private val toggleFriendsHomesSlot = InventoryCoordinate(0, getInventory.getSize / 9 - 1)
  private val showFriendsHomes = PlayerDataManager(player).get(PlayerDataManager.Attribute.SHOW_FRIEND_HOMES).toBoolean
  private val friendsHomes = if showFriendsHomes then Home.getFriendsHomes(player, realm) else List()

  private var friendSkullScrollIndex = 0

  val friends: List[String] = Friend.getFriendList(player.getUniqueId)

  private val toggleFriendsHomesSlotUpdateTimer = QuickRunnable(() =>
    val playerToShow =
      if friends.nonEmpty then
        friendSkullScrollIndex = (friendSkullScrollIndex + 1) % friends.size
        NoxetServer.getPlugin.getServer
          .getOfflinePlayer(UsernameStorageManager.getUuidFromUsernameOrUuid(
            friends.get(friendSkullScrollIndex)))
      else
        player

    setSlotItem(ItemGenerator.generatePlayerSkull(
      playerToShow,
      "§aToggle Friends' Homes",
      List(
        if showFriendsHomes then "§a§l✔ SHOWN" else "§c§l❌ HIDDEN",
        "§7When enabled, your friends' shared",
        "§7homes will be shown in this list.",
        s"§e→ Click to ${if showFriendsHomes then "hide" else "show"}."
      )
    ), toggleFriendsHomesSlot)
  ).runTaskTimer(NoxetServer.getPlugin, 0, 40)

  override protected def updateInventory(): Unit =
    for ((name, loc), i) <- homes.zipWithIndex do
      val blocksAway =
        if loc.getWorld == player.getWorld then
          Some(player.getLocation.distance(loc).toInt)
        else
          None

      val locationNote = "§5" + (loc.getWorld.getEnvironment match
        case NORMAL => "Overworld"
        case NETHER => "Nether"
        case THE_END => "The End"
        case _ => "???") +
        blocksAway
          .map(s" §e${DecimalFormat("#,###").format(blocksAway)} blocks away.")
          .getOrElse("")

      val isMainHome = name == Home.defaultHomeName

      val lore = List(
        locationNote,
        s"§8X §7${loc.getX.toInt}",
        s"§8Y §7${loc.getY.toInt}",
        s"§8Z §7${loc.getZ.toInt}",
        "§e→ Double-click to §5§nteleport§e.",
        "§e→ Right-click to §b§nrename§e.",
        "§e→ Press any number on keyboard to §c§nremove§e this home.",
        s"§e→ Shift-right-click to §d§n${if !Home.isHomeFriendShared(name) then "enable" else "disable"} friend sharing§e."
      )

      val homeItem = ItemGenerator.generateItem(
        if !isMainHome then Material.WHITE_BANNER else Material.RED_BED,
        if !Home.isHomeFriendShared(name) then
          (if !isMainHome then "§a" else "§5◆ §3") + name
        else
          "§5☮ §a" + name.substring(1),
        lore
      )

      if !isMainHome then
        generateBannerPatterns(homeItem, loc.hashCode)

      setSlotItem(homeItem, InventoryCoordinate.fromSlotIndex(i))

    for ((friend, (homeName, _)), i) <- friendsHomes.zipWithIndex do
      setSlotItem(
        ItemGenerator.generatePlayerSkull(friend.playerOption.get,
          "§3" + Home.getFriendHomeName(friend.username, homeName),
          List(
            s"§7This home is shared by §b${friend.username}§7.",
            "§e→ Double-click to §5§nteleport§e."
          )), InventoryCoordinate.fromSlotIndex(i + homes.size))

  override protected def onSlotClick(player: Player, coordinate: InventoryCoordinate, clickType: ClickType): Boolean = {
    if coordinate == toggleFriendsHomesSlot then
      PlayerDataManager(player).toggleBoolean(PlayerDataManager.Attribute.SHOW_FRIEND_HOMES).save()
      HomeNavigationMenu(player, realm).openInventory(player)
      return false

    homes.zipWithIndex // add index to...
      .find(InventoryCoordinate.fromSlotIndex(_._2) == coordinate).map(_._1) match // ... find clicked home & remove index
      case Some((name, loc)) => return clickType match
        case ClickType.DOUBLE_CLICK =>
          player.performCommand(s"home tp $name")
          true
        case ClickType.NUMBER_KEY =>
          ConfirmationMenu(s"Delete home '$name'?", () => {
            player.performCommand(s"home remove $name")
            HomeNavigationMenu(player, realm).openInventory(player)
          }, HomeNavigationMenu(player, realm).openInventory(player))
            .openInventory(player)
          true
        case ClickType.RIGHT =>
          ChatPromptMenu(
            s"new name for home '$name'",
            player,
            response =>
              player.performCommand(s"home rename $name ${response.getMessage}")
              HomeNavigationMenu(player, realm).openInventory(player)
          )
          true
        case ClickType.SHIFT_RIGHT =>
          val isAlreadyFriendShared = Home.isHomeFriendShared(name)
          ConfirmationMenu(s"${if isAlreadyFriendShared then "Disable" else "Enable"} friend sharing for home '$name'?", () =>
            player.performCommand(
              s"home rename $name " + (if isAlreadyFriendShared then name.substring(1) else s"*$name")
            )
            HomeNavigationMenu(player, realm).openInventory(player)
            , () => HomeNavigationMenu(player, realm).openInventory(player)).openInventory(player)
          true
        case _ => false
      case None => ()

    friendsHomes.zipWithIndex
      .find(InventoryCoordinate.fromSlotIndex(_._2 + homes.size) == coordinate).map(_._1) match
      case Some((player, (name, loc))) => clickType match
        case ClickType.DOUBLE_CLICK =>
          player.performCommand(s"home friend-tp ${Home.getFriendHomeName(player.username, name)}")
          true
        case _ => false
      case None => false

  private def generateBannerPatterns(banner: ItemStack, seed: Int): Unit =
    given Random(seed)

    val patterns = for _ <- 0 until 10 yield
      // Append random patterns to the banner.
      // (based on location seed)
      Pattern(DyeColor.values.sampleOne.get, PatternType.values.sampleOne)

    val bannerMeta = banner.getItemMeta.asInstanceOf[BannerMeta]
    assert(bannerMeta != null)

    bannerMeta.setPatterns(patterns)
    bannerMeta.addItemFlags(ItemFlag.HIDE_POTION_EFFECTS) // Hide banner pattern lore.

    banner.setItemMeta(bannerMeta)

  override protected def stop(): Unit =
    super.stop()
    toggleFriendsHomesSlotUpdateTimer.cancel()